File: Data Structure & Algorithm/Categories/Recursion/en/medium/226. Invert Binary Tree/226. Invert Binary Tree.ts
// 226. Invert Binary Tree

// Given the root of a binary tree, invert the tree, and return its root.

// Example 1:
// Input: root = [4,2,7,1,3,6,9]
// Output: [4,7,2,9,6,3,1]

// Example 2:
// Input: root = [2,1,3]
// Output: [2,3,1]

// Example 3:
// Input: root = []
// Output: []

// Definition for a binary tree node.
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

export function invertTree(root: TreeNode | null): TreeNode | null {
  const recursive = (node: TreeNode | null) => {
    if (!node) {
      return null;
    }

    node.left = recursive(node.left);
    node.right = recursive(node.right);

    const temp = node.left;
    node.left = node.right;
    node.right = temp;

    return node;
  };

  return recursive(root);
}

File: Data Structure & Algorithm/Source/Code Random Thoughts/2nd round/Tree/226. Invert Binary Tree/226. Invert Binary Tree.go
// 226. Invert Binary Tree

// Given the root of a binary tree, invert the tree, and return its root.

// Example 1:
// Input: root = [4,2,7,1,3,6,9]
// Output: [4,7,2,9,6,3,1]

// Example 2:
// Input: root = [2,1,3]
// Output: [2,3,1]

// Example 3:
// Input: root = []
// Output: []

package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// post order
func invertTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	var recursion func(node *TreeNode) *TreeNode
	recursion = func(node *TreeNode) *TreeNode {
		if node == nil {
			return nil
		}

		left := recursion(node.Left)
		right := recursion(node.Right)

		node.Left = right
		node.Right = left

		return node
	}

	return recursion(root)
}

// pre order
func invertTree1(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	var recursion func(node *TreeNode)
	recursion = func(node *TreeNode) {
		if node == nil {
			return
		}

		temp := node.Left
		node.Left = node.Right
		node.Right = temp

		recursion(node.Left)
		recursion(node.Right)
	}

	recursion(root)

	return root
}
